# Importing Needed Libraries

In [1]:
# import the library
%matplotlib inline

import pandas as pd
import numpy as np
import scipy
import matplotlib.pyplot as plt
from ast import literal_eval

# to give only two decimal points
pd.set_option('display.float_format', lambda x: '%.2f' % x)

# Loading Files

In [2]:
# load main file from local
df_movie = pd.read_csv(r'C:\Users\Alka.Patel\CEBD 1260\Project\Data\Raw\movies_metadata.csv',low_memory=False)
# load extra keywords file from local
df_keyword= pd.read_csv(r'C:\Users\Alka.Patel\CEBD 1260\Project\Data\Raw\keywords.csv')
# load extra rating file from local
df_rating_big= pd.read_csv(r'C:\Users\Alka.Patel\CEBD 1260\Project\Data\Raw\ratings.csv')
# load small rating file after realizing in the end, the other one won't save on my local
df_rating= pd.read_csv(r'C:\Users\Alka.Patel\CEBD 1260\Project\Data\Raw\ratings_small.csv')

I had originally tried this with the large rating file, of 26 million rows. However, the files in the end, after merging did not save on my local, so I ended up doing the acitivity with the smaller rating file of about 100 000 records from the same data source. 

# Exploring Movies

In [3]:
df_movie.head()

adult                              belongs_to_collection    budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1  False                                                NaN  65000000   
2  False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3  False                                                NaN  16000000   
4  False  {'id': 96871, 'name': 'Father of the Bride Col...         0   

                                              genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   
3  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   
4                     [{'id': 35, 'name': 'Comedy'}]   

                               homepage     id    imdb_id original_language  \
0  http://toystory.disney.com/toy-story    862  tt0114709                en   
1                                   NaN   8844  tt0113497                en   
2                                   NaN  15602  tt0113228                en   
3                                   NaN  31357  tt0114885                en   
4                                   NaN  11862  tt0113041                en   

                original_title  \
0                    Toy Story   
1                      Jumanji   
2             Grumpier Old Men   
3            Waiting to Exhale   
4  Father of the Bride Part II   

                                            overview    ...     release_date  \
0  Led by Woody, Andy's toys live happily in his ...    ...       1995-10-30   
1  When siblings Judy and Peter discover an encha...    ...       1995-12-15   
2  A family wedding reignites the ancient feud be...    ...       1995-12-22   
3  Cheated on, mistreated and stepped on, the wom...    ...       1995-12-22   
4  Just when George Banks has recovered from his ...    ...       1995-02-10   

       revenue runtime                                   spoken_languages  \
0 373554033.00   81.00           [{'iso_639_1': 'en', 'name': 'English'}]   
1 262797249.00  104.00  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
2         0.00  101.00           [{'iso_639_1': 'en', 'name': 'English'}]   
3  81452156.00  127.00           [{'iso_639_1': 'en', 'name': 'English'}]   
4  76578911.00  106.00           [{'iso_639_1': 'en', 'name': 'English'}]   

     status                                            tagline  \
0  Released                                                NaN   
1  Released          Roll the dice and unleash the excitement!   
2  Released  Still Yelling. Still Fighting. Still Ready for...   
3  Released  Friends are the people who let you be yourself...   
4  Released  Just When His World Is Back To Normal... He's ...   

                         title  video vote_average vote_count  
0                    Toy Story  False         7.70    5415.00  
1                      Jumanji  False         6.90    2413.00  
2             Grumpier Old Men  False         6.50      92.00  
3            Waiting to Exhale  False         6.10      34.00  
4  Father of the Bride Part II  False         5.70     173.00  

[5 rows x 24 columns]

In [4]:
print("Column Names")
print(df_movie.columns)
print("")
print("Table size")
print("")
print(df_movie.shape)
print("")
print("Movie Table Summary of numerical data")
print("")
print(df_movie.describe())
print("")
print("Column Data Types")
print("")
print(df_movie.dtypes)
print("")
print("Dataframe Index")
print("")
print(df_movie.index)

Column Names
Index(['adult', 'belongs_to_collection', 'budget', 'genres', 'homepage', 'id',
       'imdb_id', 'original_language', 'original_title', 'overview',
       'popularity', 'poster_path', 'production_companies',
       'production_countries', 'release_date', 'revenue', 'runtime',
       'spoken_languages', 'status', 'tagline', 'title', 'video',
       'vote_average', 'vote_count'],
      dtype='object')

Table size

(45466, 24)

Movie Table Summary of numerical data

            revenue  runtime  vote_average  vote_count
count      45460.00 45203.00      45460.00    45460.00
mean    11209348.54    94.13          5.62      109.90
std     64332246.74    38.41          1.92      491.31
min            0.00     0.00          0.00        0.00
25%            0.00    85.00          5.00        3.00
50%            0.00    95.00          6.00       10.00
75%            0.00   107.00          6.80       34.00
max   2787965087.00  1256.00         10.00    14075.00

Column Data Types

adul

# Explore Ratings Dataframe

In [5]:
df_rating.head()

userId  movieId  rating   timestamp
0       1       31    2.50  1260759144
1       1     1029    3.00  1260759179
2       1     1061    3.00  1260759182
3       1     1129    2.00  1260759185
4       1     1172    4.00  1260759205

In [6]:
print("This was the original rating table I wanted to use")
print(df_rating_big.shape)

print("__________________")
print("Smaller Rating Table Stats")
print("Column Names")
print(df_rating.columns)
print("")
print("Ratings size")
print("")
print(df_rating.shape)
print("")
print("Rating Table Summary of numerical data")
print("")
print(df_rating.describe())
print("")
print("Column Data Types")
print("")
print(df_rating.dtypes)
print("")
print("Dataframe Index")
print("")
print(df_rating.index)

This was the original rating table I wanted to use
(26024289, 4)
__________________
Smaller Rating Table Stats
Column Names
Index(['userId', 'movieId', 'rating', 'timestamp'], dtype='object')

Ratings size

(100004, 4)

Rating Table Summary of numerical data

         userId   movieId    rating     timestamp
count 100004.00 100004.00 100004.00     100004.00
mean     347.01  12548.66      3.54 1129639086.94
std      195.16  26369.20      1.06  191685826.03
min        1.00      1.00      0.50  789652009.00
25%      182.00   1028.00      3.00  965847824.00
50%      367.00   2406.50      4.00 1110421822.00
75%      520.00   5418.00      4.00 1296192495.50
max      671.00 163949.00      5.00 1476640644.00

Column Data Types

userId         int64
movieId        int64
rating       float64
timestamp      int64
dtype: object

Dataframe Index

RangeIndex(start=0, stop=100004, step=1)


# Exploring Keywords Dataframe

In [7]:
df_keyword.head()

id                                           keywords
0    862  [{'id': 931, 'name': 'jealousy'}, {'id': 4290,...
1   8844  [{'id': 10090, 'name': 'board game'}, {'id': 1...
2  15602  [{'id': 1495, 'name': 'fishing'}, {'id': 12392...
3  31357  [{'id': 818, 'name': 'based on novel'}, {'id':...
4  11862  [{'id': 1009, 'name': 'baby'}, {'id': 1599, 'n...

In [8]:
print("Column Names")
print(df_keyword.columns)
print("")
print("Ratings size")
print("")
print(df_keyword.shape)
print("")
print("Column Data Types")
print("")
print(df_keyword.dtypes)
print("")
print("Dataframe Index")
print("")
print(df_keyword.index)

Column Names
Index(['id', 'keywords'], dtype='object')

Ratings size

(46419, 2)

Column Data Types

id           int64
keywords    object
dtype: object

Dataframe Index

RangeIndex(start=0, stop=46419, step=1)


# Fixing up Movie Dataset


In [9]:
#changing name of column to match rating dataframe
df_movie = df_movie.rename(columns={'id': 'movieId'})

In [10]:
#Changing Column Data types
df_movie['release_date'] = pd.to_datetime(df_movie['release_date'], errors='coerce')
print("release date is", df_movie['release_date'].dtypes)
#Changing budjet to int
df_movie['budget'] = pd.to_numeric(df_movie['budget'], errors='coerce')
print("budget is", df_movie['budget'].dtypes)
#Changing runtime to int
df_movie['runtime'] = pd.to_numeric(df_movie['runtime'], errors='coerce')
print("runtime is", df_movie['runtime'].dtypes)
#Changing name and type of movie_id to match ratings table
df_movie['movieId'] = pd.to_numeric(df_movie['movieId'], errors='coerce').fillna(0).astype(np.int64)
print("movieID is", df_movie['movieId'].dtypes)
df_movie['popularity'] = pd.to_numeric(df_movie['popularity'], errors='coerce')
print("popularity is", df_movie['popularity'].dtypes)
#Fixing date type
df_movie['release_date']=pd.to_datetime(df_movie['release_date'])
print('release_date', df_movie['release_date'].dtypes)

release date is datetime64[ns]
budget is float64
runtime is float64
movieID is int64
popularity is float64
release_date datetime64[ns]


In [11]:
"""
Need to create a list from the Nested Genres, used this function that I got from: 
    https://www.kaggle.com/rounakbanik/movie-recommender-systems
https://stackoverflow.com/questions/52064741/extracting-json-data-from-column-in-pandas-dataframe
"""


df_movie['genres'] = df_movie['genres'].fillna('[]').apply(literal_eval).apply((lambda x: [i['name'] for i in x] if isinstance(x, list) else []))
df_movie['production_companies'] = df_movie['production_companies'].fillna('[]').apply(literal_eval).apply((lambda x: [i['name'] for i in x] if isinstance(x, list) else []))

df_movie['production_countries'] = df_movie['production_countries'].fillna('[]').apply(literal_eval).apply((lambda x: [i['name'] for i in x] if isinstance(x, list) else []))
df_movie.head()

adult                              belongs_to_collection      budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ... 30000000.00   
1  False                                                NaN 65000000.00   
2  False  {'id': 119050, 'name': 'Grumpy Old Men Collect...        0.00   
3  False                                                NaN 16000000.00   
4  False  {'id': 96871, 'name': 'Father of the Bride Col...        0.00   

                         genres                              homepage  \
0   [Animation, Comedy, Family]  http://toystory.disney.com/toy-story   
1  [Adventure, Fantasy, Family]                                   NaN   
2             [Romance, Comedy]                                   NaN   
3      [Comedy, Drama, Romance]                                   NaN   
4                      [Comedy]                                   NaN   

   movieId    imdb_id original_language               original_title  \
0      862  tt0114709                en                    Toy Story   
1     8844  tt0113497                en                      Jumanji   
2    15602  tt0113228                en             Grumpier Old Men   
3    31357  tt0114885                en            Waiting to Exhale   
4    11862  tt0113041                en  Father of the Bride Part II   

                                            overview    ...      release_date  \
0  Led by Woody, Andy's toys live happily in his ...    ...        1995-10-30   
1  When siblings Judy and Peter discover an encha...    ...        1995-12-15   
2  A family wedding reignites the ancient feud be...    ...        1995-12-22   
3  Cheated on, mistreated and stepped on, the wom...    ...        1995-12-22   
4  Just when George Banks has recovered from his ...    ...        1995-02-10   

       revenue runtime                                   spoken_languages  \
0 373554033.00   81.00           [{'iso_639_1': 'en', 'name': 'English'}]   
1 262797249.00  104.00  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
2         0.00  101.00           [{'iso_639_1': 'en', 'name': 'English'}]   
3  81452156.00  127.00           [{'iso_639_1': 'en', 'name': 'English'}]   
4  76578911.00  106.00           [{'iso_639_1': 'en', 'name': 'English'}]   

     status                                            tagline  \
0  Released                                                NaN   
1  Released          Roll the dice and unleash the excitement!   
2  Released  Still Yelling. Still Fighting. Still Ready for...   
3  Released  Friends are the people who let you be yourself...   
4  Released  Just When His World Is Back To Normal... He's ...   

                         title  video vote_average vote_count  
0                    Toy Story  False         7.70    5415.00  
1                      Jumanji  False         6.90    2413.00  
2             Grumpier Old Men  False         6.50      92.00  
3            Waiting to Exhale  False         6.10      34.00  
4  Father of the Bride Part II  False         5.70     173.00  

[5 rows x 24 columns]

# Inspecting Languages

In [12]:
print(df_movie['original_language'].value_counts())

en       32269
fr        2438
it        1529
ja        1350
de        1080
es         994
ru         826
hi         508
ko         444
zh         409
sv         384
pt         316
cn         313
fi         297
nl         248
da         225
pl         219
tr         150
cs         130
el         113
no         106
fa         101
hu         100
ta          78
th          76
he          67
sr          63
ro          57
te          45
ar          39
         ...  
kk           3
am           2
iu           2
ne           2
pa           2
mn           2
lo           2
af           2
bo           2
ps           2
fy           1
tg           1
zu           1
82.0         1
lb           1
mt           1
ay           1
eo           1
68.0         1
si           1
jv           1
rw           1
sm           1
la           1
qu           1
uz           1
cy           1
104.0        1
hy           1
gl           1
Name: original_language, Length: 92, dtype: int64


In [13]:
#Since there are many languages -92 total, I want to take only the top 10 languages as the rest are scarcely distributed

choice_languages = ['en','fr','it','ja','de','es','ru','hi','ko','zh']
df_movie= df_movie[df_movie.original_language.isin(choice_languages)]
df_movie.shape 
print(df_movie['original_language'].value_counts())

en    32269
fr     2438
it     1529
ja     1350
de     1080
es      994
ru      826
hi      508
ko      444
zh      409
Name: original_language, dtype: int64


# Selecting movies with only released as status

In [14]:
print(df_movie['status'].value_counts())

Released           41450
Rumored              200
Post Production       87
In Production         18
Planned               15
Canceled               2
Name: status, dtype: int64


In [15]:
#Here we can see that Released is 99% of all movie status, so it would be justifiable to drop all other status movies.

c=df_movie['status'].value_counts()
print(c/c.sum())


Released          0.99
Rumored           0.00
Post Production   0.00
In Production     0.00
Planned           0.00
Canceled          0.00
Name: status, dtype: float64


In [16]:
df_movie=(df_movie.loc[df_movie['status']== 'Released'])


In [17]:
df_movie.head()

adult                              belongs_to_collection      budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ... 30000000.00   
1  False                                                NaN 65000000.00   
2  False  {'id': 119050, 'name': 'Grumpy Old Men Collect...        0.00   
3  False                                                NaN 16000000.00   
4  False  {'id': 96871, 'name': 'Father of the Bride Col...        0.00   

                         genres                              homepage  \
0   [Animation, Comedy, Family]  http://toystory.disney.com/toy-story   
1  [Adventure, Fantasy, Family]                                   NaN   
2             [Romance, Comedy]                                   NaN   
3      [Comedy, Drama, Romance]                                   NaN   
4                      [Comedy]                                   NaN   

   movieId    imdb_id original_language               original_title  \
0      862  tt0114709                en                    Toy Story   
1     8844  tt0113497                en                      Jumanji   
2    15602  tt0113228                en             Grumpier Old Men   
3    31357  tt0114885                en            Waiting to Exhale   
4    11862  tt0113041                en  Father of the Bride Part II   

                                            overview    ...      release_date  \
0  Led by Woody, Andy's toys live happily in his ...    ...        1995-10-30   
1  When siblings Judy and Peter discover an encha...    ...        1995-12-15   
2  A family wedding reignites the ancient feud be...    ...        1995-12-22   
3  Cheated on, mistreated and stepped on, the wom...    ...        1995-12-22   
4  Just when George Banks has recovered from his ...    ...        1995-02-10   

       revenue runtime                                   spoken_languages  \
0 373554033.00   81.00           [{'iso_639_1': 'en', 'name': 'English'}]   
1 262797249.00  104.00  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
2         0.00  101.00           [{'iso_639_1': 'en', 'name': 'English'}]   
3  81452156.00  127.00           [{'iso_639_1': 'en', 'name': 'English'}]   
4  76578911.00  106.00           [{'iso_639_1': 'en', 'name': 'English'}]   

     status                                            tagline  \
0  Released                                                NaN   
1  Released          Roll the dice and unleash the excitement!   
2  Released  Still Yelling. Still Fighting. Still Ready for...   
3  Released  Friends are the people who let you be yourself...   
4  Released  Just When His World Is Back To Normal... He's ...   

                         title  video vote_average vote_count  
0                    Toy Story  False         7.70    5415.00  
1                      Jumanji  False         6.90    2413.00  
2             Grumpier Old Men  False         6.50      92.00  
3            Waiting to Exhale  False         6.10      34.00  
4  Father of the Bride Part II  False         5.70     173.00  

[5 rows x 24 columns]

In [18]:
'''Dropping Unnecessary Columns. I will not be doing any explorations on the below parameters. So I have decided to drop these
columns. I dropped the status column at this point, because we know all data now are only of the status, 'released'
'''
df_movie=df_movie.drop(columns=['adult','belongs_to_collection','homepage','overview','poster_path','status','tagline',
              'video','imdb_id','spoken_languages'])

# Fixing up Keyword Dataset

In [19]:
#changing name of column to match rating and movie dataframe
df_keyword = df_keyword.rename(columns={'id': 'movieId'})
df_keyword.columns

Index(['movieId', 'keywords'], dtype='object')

In [20]:
df_keyword['keywords'] = df_keyword['keywords'].fillna('[]').apply(literal_eval).apply((lambda x: [i['name'] for i in x] if isinstance(x, list) else []))
df_keyword.head()

movieId                                           keywords
0      862  [jealousy, toy, boy, friendship, friends, riva...
1     8844  [board game, disappearance, based on children'...
2    15602  [fishing, best friend, duringcreditsstinger, o...
3    31357  [based on novel, interracial relationship, sin...
4    11862  [baby, midlife crisis, confidence, aging, daug...

# Missing Values

In [21]:
# Number of missing values by columns
print(df_movie.isnull().sum())
print('')
print(df_rating.isnull().sum())
print('')
print(df_keyword.isnull().sum())

budget                    0
genres                    0
movieId                   0
original_language         0
original_title            0
popularity                0
production_companies      0
production_countries      0
release_date             70
revenue                   0
runtime                 187
title                     0
vote_average              0
vote_count                0
dtype: int64

userId       0
movieId      0
rating       0
timestamp    0
dtype: int64

movieId     0
keywords    0
dtype: int64


In [22]:
# Removing these values from dataframe as they are insignificant and status had said never released
# The percentage of release_date missing from the total dataframe is: 0.61%
rdn=df_movie['release_date'].isnull().sum()
print((rdn/df_movie['release_date'].count())*100)



0.1691638472692122


In [23]:
# dropping empty release date
df_movie= df_movie.dropna(subset=['release_date'])

In [24]:
#Similarly, 187 runtime, is insignificant considering the total from the big data set, it is only 0.4% :
rtn=df_movie['runtime'].isnull().sum()
print((rtn/df_movie['runtime'].count())*100)


0.4368932038834952


In [25]:
# dropping empty runtimes
df_movie= df_movie.dropna(subset=['runtime'])

In [26]:
print(df_movie.isnull().sum())

budget                  0
genres                  0
movieId                 0
original_language       0
original_title          0
popularity              0
production_companies    0
production_countries    0
release_date            0
revenue                 0
runtime                 0
title                   0
vote_average            0
vote_count              0
dtype: int64


# Dealing with 0 Values

In [27]:
#Curiously I want to see 0 values. 
(df_movie == 0).astype(int).sum()

budget                  32778
genres                      0
movieId                     0
original_language           0
original_title              0
popularity                 35
production_companies        0
production_countries        0
release_date                0
revenue                 34047
runtime                  1261
title                       0
vote_average             2578
vote_count               2492
dtype: int64

Originally I thought of dropping the 0s, but then I would be only left with about 7k records. So I will just have to keep
that in mind when I do any analysis on budget and revenue, that most values are 0.

In [28]:
#I want to save all my new processed tables in processed folder before merging also
df_movie.to_csv(r"C:\Users\Alka.Patel\CEBD 1260\Project\Data\Processed\movies_clean.csv", index = False, header = True)

df_keyword.to_csv(r"C:\Users\Alka.Patel\CEBD 1260\Project\Data\Processed\keyword.csv", index = False, header = True)

#The rating file is much too big to save an extra copy of, additionally no changes was done to it.

In [29]:
#Merging Dataframes
df_merge = df_movie.copy()
df_merge = df_merge.merge(df_keyword, on='movieId', how='left')
df_merge = df_merge.merge(df_rating, on='movieId', how='left')

In [30]:
df_merge.head()

budget                        genres  movieId original_language  \
0 30000000.00   [Animation, Comedy, Family]      862                en   
1 65000000.00  [Adventure, Fantasy, Family]     8844                en   
2        0.00             [Romance, Comedy]    15602                en   
3 16000000.00      [Comedy, Drama, Romance]    31357                en   
4        0.00                      [Comedy]    11862                en   

                original_title  popularity  \
0                    Toy Story       21.95   
1                      Jumanji       17.02   
2             Grumpier Old Men       11.71   
3            Waiting to Exhale        3.86   
4  Father of the Bride Part II        8.39   

                                production_companies  \
0                          [Pixar Animation Studios]   
1  [TriStar Pictures, Teitler Film, Interscope Co...   
2                     [Warner Bros., Lancaster Gate]   
3           [Twentieth Century Fox Film Corporation]   
4       [Sandollar Productions, Touchstone Pictures]   

         production_countries release_date      revenue  runtime  \
0  [United States of America]   1995-10-30 373554033.00    81.00   
1  [United States of America]   1995-12-15 262797249.00   104.00   
2  [United States of America]   1995-12-22         0.00   101.00   
3  [United States of America]   1995-12-22  81452156.00   127.00   
4  [United States of America]   1995-02-10  76578911.00   106.00   

                         title  vote_average  vote_count  \
0                    Toy Story          7.70     5415.00   
1                      Jumanji          6.90     2413.00   
2             Grumpier Old Men          6.50       92.00   
3            Waiting to Exhale          6.10       34.00   
4  Father of the Bride Part II          5.70      173.00   

                                            keywords  userId  rating  \
0  [jealousy, toy, boy, friendship, friends, riva...     nan     nan   
1  [board game, disappearance, based on children'...     nan     nan   
2  [fishing, best friend, duringcreditsstinger, o...     nan     nan   
3  [based on novel, interracial relationship, sin...     nan     nan   
4  [baby, midlife crisis, confidence, aging, daug...     nan     nan   

   timestamp  
0        nan  
1        nan  
2        nan  
3        nan  
4        nan

I need to see the null values on the merge file, because in the rating file, it has 100 000 records, but not all movie ids are there, so not all movies have a rating from grouplens, and so we have to check the null values again

In [31]:
print(df_merge.isnull().sum())

budget                      0
genres                      0
movieId                     0
original_language           0
original_title              0
popularity                  0
production_companies        0
production_countries        0
release_date                0
revenue                     0
runtime                     0
title                       0
vote_average                0
vote_count                  0
keywords                    1
userId                  39389
rating                  39389
timestamp               39389
dtype: int64


We can see that out of 100 000 ratings, a large number of ratings is null. This makes me curious and I want to do some basic exploration on the merged dataset.

In [32]:
print("Column Names")
print(df_merge.columns)
print("")
print("Merge size")
print("")
print(df_merge.shape)
print("")
print("Merge Types")
print("")
print(df_merge.dtypes)
print("")
print(df_merge.index)

Column Names
Index(['budget', 'genres', 'movieId', 'original_language', 'original_title',
       'popularity', 'production_companies', 'production_countries',
       'release_date', 'revenue', 'runtime', 'title', 'vote_average',
       'vote_count', 'keywords', 'userId', 'rating', 'timestamp'],
      dtype='object')

Merge size

(82615, 18)

Merge Types

budget                         float64
genres                          object
movieId                          int64
original_language               object
original_title                  object
popularity                     float64
production_companies            object
production_countries            object
release_date            datetime64[ns]
revenue                        float64
runtime                        float64
title                           object
vote_average                   float64
vote_count                     float64
keywords                        object
userId                         float64
rating             

In [33]:
df_merge.movieId.value_counts()

296       324
318       311
593       304
260       291
480       274
527       244
110       228
608       224
780       218
588       215
457       213
2959      202
590       202
858       200
150       200
364       200
4993      200
380       198
592       196
2762      193
2028      191
1580      190
377       180
595       176
344       175
648       168
1265      165
1721      164
3578      161
231       158
         ... 
38459       1
52784       1
9773        1
11820       1
398891      1
1577        1
32294       1
26149       1
28196       1
22051       1
231001      1
18017       1
93828       1
28260       1
87683       1
42621       1
44668       1
38523       1
65142       1
58997       1
61044       1
56946       1
52848       1
13935       1
15982       1
9837        1
11884       1
265832      1
61062       1
2049        1
Name: movieId, Length: 41177, dtype: int64

We see that with count that there are a distinct 41562 movieIds in the original table. Now how many distinct ids are there if we remove the null values from rating?


In [34]:
df_merge_ratingcleaned = df_merge.dropna(subset=['rating'])

In [35]:
#counting the movieIDs left
df_merge_ratingcleaned.movieId.value_counts()

296       324
318       311
593       304
260       291
480       274
527       244
110       228
608       224
780       218
588       215
457       213
2959      202
590       202
858       200
150       200
364       200
4993      200
380       198
592       196
2762      193
2028      191
1580      190
377       180
595       176
344       175
648       168
1265      165
1721      164
3578      161
231       158
         ... 
7340        1
25763       1
559         1
31104       1
26850       1
4173        1
8267        1
1870        1
1550        1
95744       1
5516        1
27875       1
142507      1
43899       1
3596        1
127728      1
526         1
43419       1
26947       1
2191        1
8266        1
4140        1
4108        1
1933        1
1901        1
5839        1
40574       1
1613        1
1549        1
2049        1
Name: movieId, Length: 2674, dtype: int64

In [36]:
#only distint 2674 mvoies have ratings from grouplens. I want to see this table more in detail:

print("Column Names")
print(df_merge_ratingcleaned.columns)
print("")
print("df_merge_ratingcleaned size")
print("")
print(df_merge_ratingcleaned.shape)
print("")
print("df_merge_ratingcleaned Types")
print("")
print(df_merge_ratingcleaned.dtypes)
print("")
print("df_merge_ratingcleaned Index")
print(df_merge_ratingcleaned.index)


Column Names
Index(['budget', 'genres', 'movieId', 'original_language', 'original_title',
       'popularity', 'production_companies', 'production_countries',
       'release_date', 'revenue', 'runtime', 'title', 'vote_average',
       'vote_count', 'keywords', 'userId', 'rating', 'timestamp'],
      dtype='object')

df_merge_ratingcleaned size

(43226, 18)

df_merge_ratingcleaned Types

budget                         float64
genres                          object
movieId                          int64
original_language               object
original_title                  object
popularity                     float64
production_companies            object
production_countries            object
release_date            datetime64[ns]
revenue                        float64
runtime                        float64
title                           object
vote_average                   float64
vote_count                     float64
keywords                        object
userId                  

# So if we drop the Nan values from ratings, we still have a 43226 rows left, but only of 2674 distinct movies.  However, I will save this cleaned rating merged table, in case I choose to do an analysis on the the rating from Grouplens. And I will also take in a merged table of the keywords and original movie table seperately, because that is better to do the actual movie analysis as it has data of over 40k movies - and it also has the votes from TMDB, additionally, the keyword merge would be pretty complete as it is only missing one value.

In [37]:
# Creating the keyword merged table to use in EDA
df_merge_keyword = df_movie.copy()
df_merge_keyword = df_merge_keyword.merge(df_keyword, on='movieId', how='left')

In [38]:
#dropping that one null value
df_merge_keyword= df_merge_keyword.dropna(subset=['keywords'])

In [39]:
#Confirming this table is what I wanted
print(df_merge_keyword.isnull().sum())

budget                  0
genres                  0
movieId                 0
original_language       0
original_title          0
popularity              0
production_companies    0
production_countries    0
release_date            0
revenue                 0
runtime                 0
title                   0
vote_average            0
vote_count              0
keywords                0
dtype: int64


In [40]:
print("Column Names")
print(df_merge_keyword.columns)
print("")
print("df_merge_keyword size")
print("")
print(df_merge_keyword.shape)
print("")
print("df_merge_keyword Types")
print("")
print(df_merge_keyword.dtypes)
print("")
print("df_merge_keywordIndex")
print(df_merge_keyword.index)

Column Names
Index(['budget', 'genres', 'movieId', 'original_language', 'original_title',
       'popularity', 'production_companies', 'production_countries',
       'release_date', 'revenue', 'runtime', 'title', 'vote_average',
       'vote_count', 'keywords'],
      dtype='object')

df_merge_keyword size

(42074, 15)

df_merge_keyword Types

budget                         float64
genres                          object
movieId                          int64
original_language               object
original_title                  object
popularity                     float64
production_companies            object
production_countries            object
release_date            datetime64[ns]
revenue                        float64
runtime                        float64
title                           object
vote_average                   float64
vote_count                     float64
keywords                        object
dtype: object

df_merge_keywordIndex
Int64Index([    0,     1,     2,

In [41]:
#counting the movieIDs left in this merged table !!!! SUCCESS OVER 40K records !!!!
df_merge_keyword.movieId.value_counts()

159849    6
18440     4
109962    4
105045    4
84198     4
10991     4
4912      4
11115     4
69234     4
5511      4
77221     4
15028     4
97995     4
12600     4
23305     4
99080     4
168538    4
110428    4
14788     4
152795    4
22649     4
132641    4
119916    4
45079     2
132873    2
291348    2
97350     2
39493     2
64462     2
187442    2
         ..
14834     1
53743     1
43552     1
49645     1
63978     1
55602     1
37351     1
39398     1
426469    1
26158     1
113154    1
59912     1
254473    1
335796    1
16925     1
18972     1
94747     1
31258     1
25113     1
158232    1
72214     1
533       1
68116     1
12819     1
14866     1
74257     1
10768     1
330483    1
77785     1
276895    1
Name: movieId, Length: 41176, dtype: int64

In [42]:
#Saving the dataframes for EDA
df_merge_ratingcleaned.to_csv(r"C:\Users\Alka.Patel\CEBD 1260\Project\Data\Processed\df_merge_ratingcleaned.csv", index = False, header = True)

df_merge_keyword.to_csv(r"C:\Users\Alka.Patel\CEBD 1260\Project\Data\Processed\df_merge_keyword.csv", index = False, header = True)